independent-projects/arc/runtime/src/main/java/io/quarkus/arc/LifecycleCallbackInvocationContext.java

```diff
@@ -0,0 +1,64 @@
+package io.quarkus.arc;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A simple InvocationContext implementation used for PostConstruct and PreDestroy callbacks.
+ */
+class LifecycleCallbackInvocationContext extends AbstractInvocationContext {
+
+    private volatile int position;
+
+    LifecycleCallbackInvocationContext(Constructor<?> constructor, Set<Annotation> interceptorBindings,
+            List<InterceptorInvocation> chain) {
+        super(null, null, constructor, null, null, interceptorBindings, chain);
+        this.position = 0;
+    }
+
+    LifecycleCallbackInvocationContext(Object target, Set<Annotation> interceptorBindings, List<InterceptorInvocation> chain) {
+        super(target, null, null, null, null, interceptorBindings, chain);
+        this.position = 0;
+    }
+
+    @Override
+    public Object proceed() throws Exception {
+        try {
+            if (position < chain.size()) {
+                // Invoke the next interceptor in the chain
+                invokeNext();
+            } else {
+                // The invocation of proceed in the last interceptor method in the chain
+                interceptorChainCompleted();
+            }
+            // The return value of a lifecycle callback is ignored
+            return null;
+        } catch (InvocationTargetException e) {
+            Throwable cause = e.getCause();
+            if (cause instanceof Error) {
+                throw (Error) cause;
+            }
+            if (cause instanceof Exception) {
+                throw (Exception) cause;
+            }
+            throw new RuntimeException(cause);
+        }
+    }
+
+    protected void interceptorChainCompleted() throws Exception {
+        // No-op
+    }
+
+    protected Object invokeNext() throws Exception {
+        int oldPosition = position;
+        try {
+            return chain.get(position++).invoke(this);
+        } finally {
+            position = oldPosition;
```

Comment:
Do you expect oldPosition to be different than position-- in some cases?

Reply:
That's a good question. In theory, `position--` should work.